fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/IndexMetaData.java

```diff
@@ -0,0 +1,236 @@
+ * the latest start time of an index build, and number of attempts. The class also includes additional methods to serialize
+ * ({@link #toBytes(IndexMetaData)}) / de-serialize ({@link #fromBytes(byte[])}) the index meta-data to be stored on / restored from disk.
+ */
+@API(API.Status.MAINTAINED)
+public class IndexMetaData {
+    private static final long newFormatPrefix = Long.MAX_VALUE;
+
+    @Nonnull
+    private IndexState state;
+    private long latestStartTime;
+    @Nullable
+    private FDBRecordStore.RebuildIndexReason reason;
+    private long attemptsCount;
+
+    private IndexMetaData() {
+        this.state = IndexState.READABLE;
+    }
+
+    public IndexMetaData(@Nonnull IndexState state) {
+        this.state = state;
+        this.latestStartTime = System.currentTimeMillis();
+    }
+
+    public IndexMetaData(@Nonnull IndexState state, @Nullable FDBRecordStore.RebuildIndexReason reason) {
+        this(state);
+        this.reason = reason;
+    }
+
+
+    public IndexMetaData(@Nonnull IndexMetaData otherMetaData) {
+        this.state = otherMetaData.state;
+        this.attemptsCount = otherMetaData.attemptsCount;
+        this.latestStartTime = otherMetaData.latestStartTime;
+        this.reason = otherMetaData.reason;
+    }
+
+    /**
+     * De-serialize an index meta-data.
+     * @param bytes the serialized index meta-data
+     * @return the index meta-data
+     */
+    public static IndexMetaData fromBytes(@Nonnull byte[] bytes) {
+        Tuple t = Tuple.fromBytes(bytes);
+        if (!isNewFormat(t.getLong(0))) {
+            return new IndexMetaData(IndexState.fromCode(t.get(0)));
+        } else {
+            IndexMetaData indexMetaData = new IndexMetaData();
+            for (int i = 1; i < t.size(); i += 2) {
+                switch (IndexMetaDataProto.MetaDataType.valueOf((int)t.getLong(i))) {
+                    case STATE:
+                        indexMetaData.state = IndexState.fromCode(t.get(i + 1));
+                        break;
+                    case LATEST_START_TIME:
+                        indexMetaData.latestStartTime = t.getLong(i + 1);
+                        break;
+                    case REASON:
+                        indexMetaData.reason = FDBRecordStore.RebuildIndexReason.fromCode(t.getLong(i + 1));
+                        break;
+                    case ATTEMPT_COUNT:
+                        indexMetaData.attemptsCount = t.getLong(i + 1);
+                        break;
+                    default:
+                        throw new RecordCoreStorageException("Field " + t.get(i) + " not implemented yet");
+                }
+            }
+            return indexMetaData;
+        }
+    }
+
+    /**
+     * Serialize an index meta-data.
+     * @param indexMetaData the index meta-data
+     * @return the serialized index meta-data
+     */
+    public static byte[] toBytes(IndexMetaData indexMetaData) {
```

Comment:
Nope, no reason. I can do that.

Reply:
Oh, I was assuming this would go along with a new format version, and that it would continue to use the old tuple serialization for record stores with the old version and then the new protobuf-based serialization for new record stores.

I think this has to take a new format version anyway (if it is changing the key from name to subspace key), and then the starting long can be omitted. You don't get this meta-data on existing stores until the format version upgrade, but I think that's okay?